Update server examples with scopes and collections - N1QL DML pages (#1809)
couchbase/docs-server: modules/n1ql/pages/n1ql-language-reference/insert.adoc

=== RBAC Privileges

To execute the INSERT statement, you must have the _Query Insert_ privilege on the target keyspace.

If the statement has any SELECT or RETURNING data-read clauses, then the _Query Select_ privilege is also required on the keyspaces referred in the respective clauses.
For more details about roles and privileges, see {authorization-overview}[Authorization].

> WARNING: Please note that the examples below will alter the data in your sample buckets.

Reviewer: As before, having given the warning, tell users how to mitigate it if possible
```suggestion
WARNING: Please note that the examples below will alter the data in your sample buckets.
To restore your sample data, remove and reinstall the `travel-sample` bucket.
Refer to xref:manage:manage-settings/install-sample-buckets.adoc[Sample Buckets] for details. 
```

For example,

* To execute the following statement, you must have the _Query Insert_ privilege on `pass:c[`travel-sample`.inventory.hotel]`.
+
[source,n1ql]
----
INSERT INTO `travel-sample`.inventory.hotel (KEY, VALUE)

> VALUES ("key1", { "type" : "hotel2", "name" : "new hotel scopes" });

Reviewer: Why "scopes"? This is just a single document?

INSERT INTO `travel-sample`.inventory.hotel (KEY, VALUE)
VALUES ("key1", { "type" : "hotel2", "name" : "new hotel scopes" });
----

* To execute the following statement, user must have the _Query Insert_ privilege on `pass:c[`travel-sample`]` and the _Query Select_ privilege on `pass:c[`travel-sample`]`.
* To execute the following statement, you must have the _Query Insert_ and _Query Select_ privileges on `pass:c[`travel-sample`.inventory.hotel]`.
+
[source,n1ql]
----
INSERT INTO `travel-sample` (KEY, VALUE)
VALUES ("key1", { "type" : "hotel", "name" : "new hotel" }) RETURNING *
INSERT INTO `travel-sample`.inventory.hotel (KEY, VALUE)
VALUES ("key1", { "type" : "hotel", "name" : "new hotel" }) RETURNING *;
----

* To execute the following statement, user must have the _Query Insert_ privilege on `pass:c[`travel-sample`]` and _Query Select_ privilege on `pass:c[`beer-sample`]`.
* To execute the following statement, you must have the _Query Insert_ privilege on `pass:c[`travel-sample`.inventory.hotel]` and _Query Select_ privilege on `pass:c[`beer-sample`]`.
+
[source,n1ql]
----
INSERT INTO `travel-sample` (KEY foo, VALUE bar)

> SELECT foo, bar FROM `beer-sample`

Author: Similar issue here.

The error is:
```
 "Cannot INSERT non-string key Missing field or index foo. of type value.missingValue."
 ```

Reviewer: This example never would have worked in its original form. See suggestion below

> SELECT foo, bar FROM `beer-sample`;

Reviewer: I had a similar example for the EXPLAIN command so copied the example from there. N.B. I'm using the `landmark` collection, because a brewery is more of a landmark than a hotel.
```suggestion
INSERT INTO `travel-sample`.inventory.landmark (KEY foo, VALUE bar)
SELECT META(doc).id AS foo, doc AS bar FROM `beer-sample` AS doc WHERE type = "brewery";
```

> * To execute the following statement, user must have the _Query Insert_ privilege on `pass:c[`travel-sample`.invetory.hotel]` and _Query Select_ privilege on `pass:c[`travel-sample`]`.

Reviewer: ```suggestion
* To execute the following statement, user must have the _Query Insert_ privilege on `pass:c[`travel-sample`.inventory.hotel]` and _Query Select_ privilege on `pass:c[`travel-sample`]`.
```

INSERT INTO `travel-sample`.inventory.hotel (KEY foo, VALUE bar)
SELECT foo, bar FROM `beer-sample`;
----

* To execute the following statement, user must have the _Query Insert_ privilege on `pass:c[`travel-sample`]` and _Query Select_ privilege on `pass:c[`travel-sample`]`.
* To execute the following statement, user must have the _Query Insert_ privilege on `pass:c[`travel-sample`.invetory.hotel]` and _Query Select_ privilege on `pass:c[`travel-sample`]`.
+
[source,n1ql]
----
INSERT INTO `travel-sample` (KEY foo, VALUE bar)

> SELECT "foo" || meta().id, bar FROM `travel-sample` WHERE type = "hotel"

Author: Same here.

```
Cannot INSERT non-string key Missing field or index foo. of type value.missingValue.
```

Reviewer: Same as before, this would never have worked in its original form. I've made a suggestion based on my suggestions for lines 95-96, but haven't tested it. Please can you run it and make sure it works?

> SELECT "foo" || meta().id, bar FROM `travel-sample`.inventory.hotel;

Reviewer: ```suggestion
INSERT INTO `travel-sample`.inventory.hotel (KEY foo, VALUE bar)
SELECT "copy_" || meta(doc).id AS foo, doc AS bar FROM `travel-sample`.inventory.hotel AS doc;
```

To restore your sample data, remove and reinstall the `travel-sample` bucket.
Refer to xref:manage:manage-settings/install-sample-buckets.adoc[Sample Buckets] for details.

For example,

* To execute the following statement, you must have the _Query Insert_ privilege on `pass:c[`travel-sample`.inventory.hotel]`.
+
[source,n1ql]
----
INSERT INTO `travel-sample`.inventory.hotel (KEY, VALUE)
VALUES ("key1", { "type" : "hotel", "name" : "new hotel" });
----

* To execute the following statement, you must have the _Query Insert_ and _Query Select_ privileges on `pass:c[`travel-sample`.inventory.hotel]`.
+
[source,n1ql]
----
INSERT INTO `travel-sample`.inventory.hotel (KEY, VALUE)
VALUES ("key1", { "type" : "hotel", "name" : "new hotel" }) RETURNING *;
----

* To execute the following statement, you must have the _Query Insert_ privilege on `pass:c[`travel-sample`.inventory.hotel]` and _Query Select_ privilege on `pass:c[`beer-sample`]`.
+
[source,n1ql]
----
INSERT INTO `travel-sample`.inventory.landmark (KEY foo, VALUE bar)
SELECT META(doc).id AS foo, doc AS bar FROM `beer-sample` AS doc WHERE type = "brewery";
----

> * To execute the following statement, you must have the _Query Insert_ privilege on `pass:c[`travel-sample`.inventory.hotel]` and _Query Select_ privilege on `pass:c[`travel-sample`]`.

Reviewer: ```suggestion
* To execute the following statement, you must have the _Query Insert_ and _Query Select_ privileges on `pass:c[`travel-sample`.inventory.hotel]`.
```

> .Query

Reviewer: In description for example-7a (line 365) change travel-sample to airline:
```
Insert a document into the `airline` keyspace using an expiration of 5 days.
```

> .Query

Reviewer: Update the description for example-8 (line 420):
```
Query the `airport` keyspace for documents with `airportname` "Heathrow", and then insert the projection (1 document) into the `airport` keyspace using a unique key generated using `UUID()`.
```

> .Query

Reviewer: Update the description for example-8a (line 459):
```
Query the `airport` keyspace for documents with `airportname` "Heathrow", and then insert the projection into the `airport` keyspace using a unique key and an expiration of 2 hours.
```

.Query
[source,n1ql]
----
INSERT INTO `travel-sample`.inventory.airport (KEY UUID(), VALUE doc, OPTIONS {"expiration": ttl})

>     SELECT META(a).expiration AS ttl, a AS doc FROM `travel-sample` a

Reviewer: Missed this one
```suggestion
    SELECT META(a).expiration AS ttl, a AS doc FROM `travel-sample`.inventory.airport a
```

> .Query

Reviewer: Update the description for example-1 (line 667)
```
The following statement inserts a single JSON document into the `airline` keyspace with key "k001".
```

> .Query

Reviewer: Update the description for example-11 (line 718)
```
Insert a new document with `key` "1025" into the `airline` keyspace.
```

> .Query

Reviewer: Update the description for example-12 (line 776)
```
Insert two documents with `key` "airline_4444" and "airline_4445" into the `airline` keyspace:
```

> .Query

Reviewer: Update the description for example-13 (lines 851-852):
```
Query the `airport` keyspace for documents where the `airportname` is "Heathrow".
Then insert the result of the select statement (a well-formed JSON document) into the `airport` keyspace with a key generated using the `UUID()` function.
```

[source,console]
----

> $ curl -v http://localhost:8093/query/service -u Administrator:password -d 'prepared="ins_india"&$key="airport_10002" & $faa_code="BLR"'

Author: I've not been able to reproduce the error we're expecting here when I run the curl command, so not 100% how to cover this. The request  seems to succeed for me even with the spacing before the '&'.

Reviewer: Yep, works fine for me too. Delete that entire Note block.

.Query
[source,n1ql]
----

> EXPLAIN INSERT INTO `travel-sample`.inventory.airport (key UUID(), value airport)
>     SELECT airport FROM `travel-sample`
>       WHERE type = "airport" AND airportname = "Heathrow";

Reviewer: I think the source should be the airport collection also. So effectively this just creates a copy of the Heathrow record. Unfortunately this change means you'll have to update the results of this query again, sorry about that.
```suggestion
EXPLAIN INSERT INTO `travel-sample`.inventory.airport (key UUID(), value airport)
    SELECT airport FROM `travel-sample`.inventory.airport
      WHERE airportname = "Heathrow";
```

          "#operator": "Sequence",
          "~children": [
            {

>               "#operator": "IntersectScan",

Reviewer: Make sure to add the callout numbers back into the examples when you update them
```suggestion
              "#operator": "IntersectScan", // <3>
```

              "scans": [
                {

>                   "#operator": "IndexScan3",

Reviewer: Make sure to add the callout numbers back into the examples when you update them
```suggestion
                  "#operator": "IndexScan3", // <2>
```

                      "exact": true,
                      "range": [
                        {
                          "high": "\"airport\"",
                          "inclusion": 3,
                          "low": "\"airport\""
                        }
                      ]
                    }
                  ],
                  "using": "gsi"
                },
                {

>                   "#operator": "IndexScan3",

Reviewer: Make sure to add the callout numbers back into the examples when you update them
```suggestion
                  "#operator": "IndexScan3", // <1>
```

              ]
            },
            {

>               "#operator": "Fetch",

Reviewer: Make sure to add the callout numbers back into the examples when you update them
```suggestion
              "#operator": "Fetch", // <4>
```

        }
      ]
    },
    "text": "INSERT INTO `travel-sample`.inventory.airport (key UUID(), value airport)\n    SELECT airport FROM `travel-sample`.inventory.airport\n      WHERE airportname = \"Heathrow\";"
  }
]
----

The Query Engine first executes the `SELECT` statement and then uses the projection to insert into the `travel-sample`.inventory.airport keyspace, performing the operations in the order listed:

<1> An `IndexScan` to search for documents using the `def_inventory_airport_airportname` index.
<2> A `Fetch` for the document in the `travel-sample`.inventory.airport keyspace.
<3> A `Filter` for documents with `airportname="Heathrow"`.

> <4> An `Insert` of the value along with the auto-generated key into the `travel-sample`.inventory.airport keyspace.

Reviewer: ```suggestion
<2> A `Fetch` for the document in the `airport` keyspace.
<3> A `Filter` for documents with `airportname="Heathrow"`.
<4> An `Insert` of the value along with the auto-generated key into the `airport` keyspace.
```

        }
      ]
    },
    "text": "INSERT INTO `travel-sample`.inventory.airport (key UUID(), value airport)\n    SELECT airport FROM `travel-sample`.inventory.airport\n      WHERE airportname = \"Heathrow\";"
  }
]
----

The Query Engine first executes the `SELECT` statement and then uses the projection to insert into the `travel-sample`.inventory.airport keyspace, performing the operations in the order listed:

> <1> An `IndexScan` to search for documents using the `def_inventory_airport_airportname` index.

Author: Given the output is now a little bit different, since we're changing the query, I am not sure if I've explained each point correctly, happy to take some suggestions here! 😓

Reviewer: Looks good, just some minor suggestions — see below.
